pytorch_lightning/metrics/classification/utils.py

```diff
@@ -0,0 +1,390 @@
+
+    In multi-class case, normally both preds and targets become ``(N, C)`` binary tensors; targets
+    by a one-hot transformation and preds by selecting ``top_k`` largest entries (if their original
+    shape was ``(N,C)``). However, if ``is_multiclass=False``, then targets and preds will be
+    returned as ``(N,1)`` tensor.
+
+    In multi-label case, normally targets and preds are returned as ``(N, C)`` binary tensors, with
+    preds being binarized as in the binary case. Here the ``C`` dimension is obtained by flattening
+    all dimensions after the first one. However if ``is_multiclass=True``, then both are returned as
+    ``(N, 2, C)``, by an equivalent transformation as in the binary case.
+
+    In multi-dimensional multi-class case, normally both target and preds are returned as
+    ``(N, C, X)`` tensors, with ``X`` resulting from flattening of all dimensions except ``N`` and
+    ``C``. The transformations performed here are equivalent to the multi-class case. However, if
+    ``is_multiclass=False`` (and there are up to two classes), then the data is returned as
+    ``(N, X)`` binary tensors (multi-label).
+
+    Also, in multi-dimensional multi-class case, if the position of the ``C``
+    dimension is ambiguous (e.g. if targets are a ``(7, 3)`` tensor, while predictions are a
+    ``(7, 3, 3)`` tensor), it will be assumed that the ``C`` dimension is the second dimension.
+    If this is not the case,  you should move it from the last to second place using
+    ``torch.movedim(preds, -1, 1)``, or using ``preds.permute``, if you are using an older
+    version of Pytorch.
+
+    Note that where a one-hot transformation needs to be performed and the number of classes
+    is not implicitly given by a ``C`` dimension, the new ``C`` dimension will either be
+    equal to ``num_classes``, if it is given, or the maximum label value in preds and
+    target.
+
+    Args:
+        preds: tensor with predictions
+        target: tensor with ground truth labels, always integers
+        threshold:
+            Threshold probability value for transforming probability predictions to binary
+            (0,1) predictions, in the case of binary or multi-label inputs. Default: 0.5
+        num_classes: number of classes
+        top_k: number of highest probability entries for each sample to convert to 1s, relevant
+            only for (multi-dimensional) multi-class cases.
+        is_multiclass: if True, treat binary and multi-label inputs as multi-class or multi-dim
+            multi-class with 2 classes, respectively. If False, treat multi-class and multi-dim
+            multi-class inputs with 1 or 2 classes as binary and multi-label, respectively.
+            Defaults to None, which treats inputs as they appear.
+
+    Returns:
+        preds: binary tensor of shape (N, C) or (N, C, X)
+        target: binary tensor of shape (N, C) or (N, C, X)
+    """
+    preds, target = preds.clone().detach(), target.clone().detach()
```

Comment:
But for classification metrics at least (which this function is for), would there even be any way to back-propagate? With the select-top-k and thresholding that is being done.

Reply:
https://discuss.pytorch.org/t/backward-through-topk-operation-on-variable/9197/2

Reply:
I see. Is there any possibility that would cause some graphs not to be destroyed between batches if this is not detached? Or does the `Metric` class handle state updates by detaching?

Reply:
It is described at the bottom of this section https://pytorch-lightning.readthedocs.io/en/stable/metrics.html#implementing-a-metric in the docs. Basically, right now you can backprop the current batch if you use forward:
```
val = metric(pred, target) # this value can be backpropagated
```
but not the accumulated metric value
```
val = metric.compute() # this value cannot be backpropagated
```

Reply:
Sorry, my question was more about the following:
Say you update the state (within a metric), so that
```
self.state += non_detached_tensor
```
Then, if I am not wrong, `self.state` also acquires `requires_grad=True` and is included in the computation graph. So is this state than detached, and if so, when does this happen?

Reply:
As the no grad happens outside `update`, everything inside should already have `requires_grad=False` so it should not be a problem I think
https://github.com/PyTorchLightning/pytorch-lightning/blob/f2fa3c82567ade0aa5bea3aa298542fc5040e4b7/pytorch_lightning/metrics/metric.py#L155-L156